File: LeetCodeAlgos/invertBinaryTree.py
# Given the root of a binary tree, invert the tree, and return its root.


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: return None
        first, next = [root], []
        while first:
            for x in first:
                if x.left:
                    next.append(x.left)
                if x.right:
                    next.append(x.right)
                x.left, x.right = x.right, x.left
            first, next = next, []
        return root

s = Solution()
# print(s.invertTree(TreeNode( 4, TreeNode( 2, TreeNode( 1 ), TreeNode( 3 )), TreeNode( 7, TreeNode( 6 ), TreeNode( 9, )))));

# print(s.invertTree(TreeNode( 2, TreeNode( 1 ), TreeNode( 3 ))));

print(s.invertTree(TreeNode( 1, None, TreeNode( 2 ))));

# print(s.invertTree([]));

File: LeetCodeAlgos/Java/InvertBinaryTree.java
// Given the root of a binary tree, invert the tree, and return its root.

import java.util.*;

public class InvertBinaryTree {
  public static void main(String[] args) {
    Solution s = new Solution();

    TreeNode tree1 = new TreeNode(4,
      new TreeNode(2,
        new TreeNode(1),
        new TreeNode(3)
      ),
      new TreeNode(7,
        new TreeNode(6),
        new TreeNode(9)
      )
    );
    TreeNode expected1 = new TreeNode(4,
      new TreeNode(7,
        new TreeNode(6),
        new TreeNode(9)
      ),
      new TreeNode(2,
        new TreeNode(1),
        new TreeNode(3)
      )
    );
    testInvertTree(1, tree1, expected1, s);

    TreeNode tree2 = new TreeNode(2,
      new TreeNode(1),
      new TreeNode(3)
    );
    TreeNode expected2 = new TreeNode(2,
      new TreeNode(3),
      new TreeNode(1)
    );
    testInvertTree(2, tree2, expected2, s);
  }

  static void testInvertTree(int testNum, TreeNode root, TreeNode expected, Solution s){
    TreeNode result = s.invertTree(root);

    System.out.println(String.format(
      "Test %d: %b",
      testNum,
      result.equals(expected)
    ));
  }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public boolean equals(TreeNode other){
      return this.val == other.val && this.left.equals(other.left) && this.right.equals(other.right);
    }
}

class Solution {
  public TreeNode invertTree(TreeNode root) {
    if (root == null) return root;

    Queue<TreeNode> nxt = new LinkedList<TreeNode>();
    nxt.add(root);

    while (!nxt.isEmpty()){
      TreeNode node = nxt.poll();
      if (node.left == null && node.right == null){
        continue;
      }
      if (node.left != null){
        nxt.add(node.left);
      }
      if (node.right!= null){
        nxt.add(node.right);
      }
      TreeNode temp = node.left;
      node.left = node.right;
      node.right = temp;
    }
    return root;
  }
}